ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,534 @@
+              account = itr.next();
+              // @todo check if remoteAccount with the same id exists. If so, check
+              // @todo account.getGenerationId() == remoteAccount.getGenerationId()+1, or fail the update operation.
+              accountMap.put(String.valueOf(account.getId()), account.toJson().toString());
+            }
+          } catch (Exception e) {
+            String message = new StringBuilder(
+                "Updating accounts failed because unexpected exception occurred when updating accountId=").append(
+                account.getId()).append(" accountName=").append(account.getName()).toString();
+            // Do not depend on Helix to log, so log the error message here.
+            logger.error(message, e);
+            throw new RuntimeException(message, e);
+          }
+          newRecord.setMapField(ACCOUNT_METADATA_MAP_KEY, accountMap);
+          return newRecord;
+        }
+      }, AccessOption.PERSISTENT);
+    }
+    if (hasSucceeded) {
+      logger.trace("Completed updating accounts, took time={}ms", System.currentTimeMillis() - startTimeMs);
+    } else {
+      logger.debug("Failed updating accounts, took time={}ms", System.currentTimeMillis() - startTimeMs);
+    }
+    return hasSucceeded;
+  }
+
+  @Override
+  public void onMessage(String topic, String message) {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+    long startTimeMs = System.currentTimeMillis();
+    logger.trace("Start to process message={} for topic={}", message, topic);
+    try {
+      switch (message) {
+        case FULL_ACCOUNT_METADATA_CHANGE_MESSAGE:
+          readFullAccountAndUpdateCache(FULL_ACCOUNT_METADATA_PATH);
+          break;
+
+        default:
+          throw new IllegalArgumentException("Could not understand message=" + message + " for topic=" + topic);
+      }
+    } catch (Exception e) {
+      logger.error("Failed to process message={} for topic={}.", message, topic, e);
+    }
+    logger.trace("Completed processing message={} for topic={}, took time={}ms", message, topic,
+        System.currentTimeMillis() - startTimeMs);
+  }
+
+  @Override
+  public void close() {
+    try {
+      isOpen = false;
+      cache.close();
+      helixStore.stop();
+    } catch (Exception e) {
+      logger.error("Exception occurred when closing HelixAccountService.", e);
+    }
+  }
+
+  /**
+   * Reads the full set of {@link Account} metadata from {@link HelixPropertyStore}, and update the local cache.
+   *
+   * @param pathToFullAccountMetadata The path to read the full set of {@link Account} metadata.
+   */
+  private void readFullAccountAndUpdateCache(String pathToFullAccountMetadata) {
+    long startTimeMs = System.currentTimeMillis();
+    logger.trace("Start fetching full account metadata set from path={}", pathToFullAccountMetadata);
+    ZNRecord zNRecord = helixStore.get(pathToFullAccountMetadata, null, AccessOption.PERSISTENT);
```

Comment:
The second arg does not really mean too much for our use case. It is simply a confirmation saying the record should be `AccessOption.PERSISTENT`. If that does not match, I expect the operation to fail.